libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -110,70 +134,148 @@ public static bool IsTrustedServiceUrl(string serviceUrl)
+                            }
+                        });
+                }
+            }
+        }
+
+        /// <summary>
+        /// Do not call this except from GetTokenRefreshAsync() 
+        /// </summary>
+        /// <returns></returns>
+        private Task<OAuthResponse> _getCurrentTokenTask(bool forceRefresh)
+        {
+            Task<OAuthResponse> oAuthTokenTask;
+
+            // if there is not a task or we are forcing it
+            if (forceRefresh || tokenTaskCache.TryGetValue(TokenCacheKey, out oAuthTokenTask) == false)
             {
-                _cache[TokenCacheKey] = token;
+                // create it
+                oAuthTokenTask = RefreshTokenAsync();
+                tokenTaskCache[TokenCacheKey] = oAuthTokenTask;
+
+                // set initial refresh time
+                autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + AutoTokenRefreshTimeSpan;
+            }
+            // if task is in faulted or canceled state then replace it with another attempt
+            else if (oAuthTokenTask.IsFaulted || oAuthTokenTask.IsCanceled)
+            {
+                oAuthTokenTask = RefreshTokenAsync();
+                tokenTaskCache[TokenCacheKey] = oAuthTokenTask;
+
+                // set initial refresh time
+                autoRefreshTimes[TokenCacheKey] = DateTime.UtcNow + AutoTokenRefreshTimeSpan;
             }
 
-            return token.access_token;
+            return oAuthTokenTask;
         }
 
         private bool ShouldSetToken(HttpRequestMessage request)
         {
-            if (IsTrustedUrl(request.RequestUri))
+            if (TrustedUri(request.RequestUri))
             {
                 return true;
             }
 
+            System.Diagnostics.Debug.WriteLine($"Service url {request.RequestUri.Authority} is not trusted and JwtToken cannot be sent to it.");
             return false;
         }
 
-        private static bool IsTrustedUrl(Uri uri)
+        private static bool TrustedUri(Uri uri)
         {
-            lock (_trustedHostNamesSync)
+            DateTime trustedServiceUrlExpiration;
+            if (TrustedHostNames.TryGetValue(uri.Host, out trustedServiceUrlExpiration))
             {
-                if (_trustedHostNames.TryGetValue(uri.Host, out DateTime trustedServiceUrlExpiration))
+                // check if the trusted service url is still valid
+                if (trustedServiceUrlExpiration > DateTime.UtcNow.Subtract(TimeSpan.FromMinutes(5)))
                 {
-                    // check if the trusted service url is still valid
-                    if (trustedServiceUrlExpiration > DateTime.UtcNow.Subtract(TimeSpan.FromMinutes(5)))
-                    {
-                        return true;
-                    }
+                    return true;
                 }
-                return false;
             }
+            return false;
         }
-
+        
         public sealed class OAuthException : Exception
```

Comment:
Do we need this class? Seems like something that could be removed.

Reply:
... and if we need it, consider moving it to it's own file. We have a fairly well followed "1 class per file" policy.

Reply:
Hmm...this would be a breaking change from V3 behavior to remove and to put in own file would not have it be a internal public class which could break the signature of it.

I think we have to live with it.

---
In reply to: [198678243](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198678243) [](ancestors = 198678243)